### DISCUSSION TITLE
Grouping or extending the information on a Scenario

### DISCUSSION THREAD

**Discussion by freddybushboy** [kenneth-gray/scenario-mock-server#11] [Ideas]
I partly bring this up because, as someone currently using [data-mocks-server](https://github.com/kenneth-gray/data-mocks-server), it will make the migration path a little easier (we use the "group" property that isn't in this version) but also as we have so many scenarios that its helpful for us to be able to group them.

I can see why `group` wasn't brought across as only one scenario can be selected now and one of the main reasons for group was to enforce only one scenario from each group was selected, but I do think being able to display related scenarios together in the selection UI is useful. It seems like adding group and displaying the radios for each group together with an additional label would we a fairly simply solution.

Another way to come at it (keeping in mind that its possible for people to create their own selection ui using the scenarios endpoint) would be to allow the addition of custom properties on the scenarios, that way people have more flexibility, could add more complex properties like `tags` and so on, but this library wouldn't need to worry about it.

Thoughts?

**Comment by kenneth-gray** [comment]
> allow the addition of custom properties on the scenarios

I think this is a good idea. Currently the response shape:

```ts
type Response = Array<{
  id: string;
  name: string;
  description: null | string;
  selected: boolean;
}>;
```

I'd suggest adding a new property called `custom` to the individual object (in case this library adds something else).

I'm not sure whether `custom` should be limited in what it allows though. Should it be:

```ts
type CustomProp = Record<string, JSON>;
```

or

```ts
type CustomProp = JSON;
```

**Reply by freddybushboy** [reply]
Not familiar with how `JSON` works as a type - does that mean it needs to be a string? You're wanting to limit it to only have `string | number | boolean` values? Plus arrays and objects I guess... You could just allow `any` but throw an error if it can't be parsed to JSON?

`JSON.parse()` returns an `any` ultimately, so I wonder if its worth it?

**Reply by kenneth-gray** [reply]
Ah yeah, it was just an example for `JSON` - I have no idea if there is a type that can be used here. But yeah `unknown` will do I guess.

**Comment by kenneth-gray** [comment]
> It seems like adding group and displaying the radios for each group together with an additional label would we a fairly simply solution

Yep, simple enough. Although needs thinking through whether the shape of the response should change so it groups by group (which is what it did originally). Alternatively it could just add a new property to the object and then worry about grouping in the UI directly (probably my preferred solution if this was needed).

**Reply by kenneth-gray** [reply]
@freddybushboy Have you got a suggestion on how it would all be declared? The scenario itself could just reference `group: '...'`, but for naming the scenario in the UI, what would you do? I believe originally it would just capitalise the name of the group for each scenario, but that's a bit hacky.

**Reply by freddybushboy** [reply]
Ah yeah good shout, I definitely wouldn't want to change the response to actually do any grouping so I think we are aligned there, and agree the capitalising is a bit hacky and looks crap if you have a group that uses an initialism etc.

I think changing the structure too much for this feature doesn't seem worth it - you can allow for an additional optional property that gets passed to run called `groupInfo?: { id: string; label: string ;}[]` or something though that also feels a bit inconsistent - scenarios aren't labelled that way.

**Reply by kenneth-gray** [reply]
To make it consistent with scenarios:

```
{
  groups: Record<string, string>
}
```

where key is the id and value is the name/label.

**Reply by kenneth-gray** [reply]
https://github.com/kenneth-gray/scenario-mock-server/pull/12

**Reply by kenneth-gray** [reply]
Groups functionality now published in 1.2.0 :tada:
